### PR TITLE
Revert fix for issue #1706

### PR DESCRIPTION
**Pull request by songrx1997**
/assign @swetharepakula 
cc @bowei @sawsa307 @gauravkghildiyal

### CODE REVIEW

**Comment by gauravkghildiyal** [comment]
This /lgtm. Thanks @songrx1997 

Some thoughts:

I think we may still want to rethink about doing this. Reverting this again opens us to situations where we explicitly allow out of band changes. In the future, the code could evolve in such a way that we unknowingly delete the security policy (maybe through another code path) and that could once again be troublesome.

File: pkg/backends/features/securitypolicy.go [hunk 1]
```diff
@@ -31,30 +31,27 @@ import (
 // EnsureSecurityPolicy ensures the security policy link on backend service.
 // TODO(mrhohn): Emit event when attach/detach security policy to backend service.
 func EnsureSecurityPolicy(cloud *gce.Cloud, sp utils.ServicePort, be *composite.BackendService) error {
-	var desiredPolicyName string
-	if sp.BackendConfig.Spec.SecurityPolicy != nil {
-		desiredPolicyName = sp.BackendConfig.Spec.SecurityPolicy.Name
-	} else {
-		desiredPolicyName = ""
+	if sp.BackendConfig.Spec.SecurityPolicy == nil {
+		return nil
```
Comment (swetharepakula) [inline pkg/backends/features/securitypolicy.go:35]: log that we see an empty security policy

Comment (bowei) [reply pkg/backends/features/securitypolicy.go:35]: or == ""

Comment (bowei) [reply pkg/backends/features/securitypolicy.go:35]: Please add the logging.

Comment (bowei) [reply pkg/backends/features/securitypolicy.go:35]: klog.V(2).Infof("Ignoring empty SecurityPolicy on backend service %s (%s:%s)", be.Name, sp.ID.Service, sp.ID.Port)

Comment (songrx1997) [reply pkg/backends/features/securitypolicy.go:36]: Updated

File: pkg/backends/features/securitypolicy.go [hunk 2]
```diff
@@ -31,30 +31,28 @@ import (
...
 
 	existingPolicyName, err := utils.KeyName(be.SecurityPolicy)
 	// The parser returns error for empty values.
 	if be.SecurityPolicy != "" && err != nil {
 		return err
 	}
-
+	desiredPolicyName := sp.BackendConfig.Spec.SecurityPolicy.Name
 	if existingPolicyName == desiredPolicyName {
 		return nil
```
Comment (bowei) [inline pkg/backends/features/securitypolicy.go:50]: klog.V(2).Infof("SecurityPolicy on backend service is not changed %s (%s:%s): %q", be.Name, sp.ID.Service, sp.ID.Port, sp.BackendConfig.Spec.SecurityPolicy.Name)

File: pkg/backends/features/securitypolicy.go [hunk 3]
```diff
@@ -31,30 +31,28 @@ import (
...
 	}
 
-	if be.Scope != meta.Global {
-		return fmt.Errorf("cloud armor security policies not supported for %s backend service %s", be.Scope, be.Name)
-	}
-
-	klog.V(2).Infof("Set security policy in backend service %s (%s:%s) to %q", be.Name, sp.ID.Service.String(), sp.ID.Port.String(), desiredPolicyName)
+	klog.V(2).Infof("Set security policy in backend service %s (%s:%s) from %s to %s", be.Name, sp.ID.Service.String(), sp.ID.Port.String(), existingPolicyName, desiredPolicyName)
```
Comment (bowei) [inline pkg/backends/features/securitypolicy.go:53]: Use the quoted string

klog.V(2).Infof("Set security policy in backend service %s (%s:%s) from %q to %q", be.Name, sp.ID.Service.String(), sp.ID.Port.String(), existingPolicyName, desiredPolicyName)

File: pkg/backends/features/securitypolicy_test.go [hunk 4]
```diff
@@ -78,7 +78,7 @@ func TestEnsureSecurityPolicy(t *testing.T) {
 			expectSetCall: true,
 		},
 		{
-			desc: "unset-policy-empty-policy-string",
+			desc: "remove-policy",
```
Comment (bowei) [inline pkg/backends/features/securitypolicy_test.go:81]: Don't we need to document that this test case is to test that we don't accidentally overwrite the user's configuration.

File: pkg/backends/features/securitypolicy.go [hunk 5]
```diff
@@ -31,30 +31,28 @@ import (
...
 // TODO(mrhohn): Emit event when attach/detach security policy to backend service.
 func EnsureSecurityPolicy(cloud *gce.Cloud, sp utils.ServicePort, be *composite.BackendService) error {
-	var desiredPolicyName string
-	if sp.BackendConfig.Spec.SecurityPolicy != nil {
-		desiredPolicyName = sp.BackendConfig.Spec.SecurityPolicy.Name
-	} else {
-		desiredPolicyName = ""
+	if sp.BackendConfig.Spec.SecurityPolicy == nil || sp.BackendConfig.Spec.SecurityPolicy.Name == "" {
+		klog.V(2).Infof("Ignoring empty SecurityPolicy on backend service %s (%s:%s)", be.Name, sp.ID.Service, sp.ID.Port)
+		return nil
```
Comment (bowei) [inline pkg/backends/features/securitypolicy.go:36]: We should put a comment here to record WHY this choice is being made

```
// Treat empty SecurityPolicy as ignored. It is too dangerous to remove user's
// security policy that may have been configured via the UI or gcloud directly
// rather than via the Kubernetes API.
```

File: pkg/backends/features/securitypolicy.go (hunk 1 above)
Comment (songrx1997) [reply pkg/backends/features/securitypolicy.go:36]: Have a question here:
How do we handle cases like this:
```
apiVersion: cloud.google.com/v1
kind: BackendConfig
metadata:
  name: my-backendconfig
spec:
  healthCheck:
    checkIntervalSec: 20
    timeoutSec: 15
  securityPolicy:
    name:
```
Shall we ignore it or not?

Comment (gauravkghildiyal) [reply pkg/backends/features/securitypolicy.go:39]: I would argue we not add `|| *sp.BackendConfig.Spec.SecurityPolicy == ""`. 

1. Because otherwise, there is no way to ever remove a security policy from within GKE.
2. Keeps us inline with the prior behaviour

Comment (bowei) [reply pkg/backends/features/securitypolicy.go:39]: I think they can remove it if they specify "None" -- can someone try this out?

Comment (bowei) [reply pkg/backends/features/securitypolicy.go:39]: oh I see -- is this what the prior code did?

Comment (bowei) [reply pkg/backends/features/securitypolicy.go:39]: Yes, let's stick to the old behavior.

Comment (songrx1997) [reply pkg/backends/features/securitypolicy.go:39]: I think so. If we set it to null. This will set it to ""

Comment (bowei) [reply pkg/backends/features/securitypolicy.go:39]: remove the || ....

File: pkg/backends/features/securitypolicy_test.go (hunk 4 above)
Comment (songrx1997) [reply pkg/backends/features/securitypolicy_test.go:81]: Since we keep the previous logic, will not update this test.

File: pkg/backends/features/securitypolicy.go (hunk 2 above)
Comment (songrx1997) [reply pkg/backends/features/securitypolicy.go:55]: Done

File: pkg/backends/features/securitypolicy.go (hunk 3 above)
Comment (songrx1997) [reply pkg/backends/features/securitypolicy.go]: Done

File: pkg/backends/features/securitypolicy.go (hunk 5 above)
Comment (songrx1997) [reply pkg/backends/features/securitypolicy.go:39]: Done

**Comment by swetharepakula** [comment]
Can we add some metrics about cloud armor attachments and detachments as part of this?

Possibly metrics of when we see a discrepancy in the configs?

File: pkg/backends/features/securitypolicy.go [hunk 6]
```diff
@@ -44,17 +49,14 @@ func EnsureSecurityPolicy(cloud *gce.Cloud, sp utils.ServicePort, be *composite.
...
-		return fmt.Errorf("cloud armor security policies not supported for %s backend service %s", be.Scope, be.Name)
-	}
-
-	klog.V(2).Infof("Set security policy in backend service %s (%s:%s) to %q", be.Name, sp.ID.Service.String(), sp.ID.Port.String(), desiredPolicyName)
+	klog.V(2).Infof("Set security policy in backend service %s (%s:%s) from %q to %q", be.Name, sp.ID.Service.String(), sp.ID.Port.String(), existingPolicyName, desiredPolicyName)
 	if err := composite.SetSecurityPolicy(cloud, be, desiredPolicyName); err != nil {
-		return fmt.Errorf("failed to set security policy %q for backend service %s (%s:%s): %v", desiredPolicyName, be.Name, sp.ID.Service.String(), sp.ID.Port.String(), err)
+		return fmt.Errorf("failed to set security policy from %q to %q for backend service %s (%s:%s): %v", existingPolicyName, desiredPolicyName, be.Name, sp.ID.Service.String(), sp.ID.Port.String(), err)
 	}
 	return nil
```
Comment (swetharepakula) [inline pkg/backends/features/securitypolicy.go:61]: add log line for completing the operation

File: pkg/backends/features/securitypolicy_test.go (hunk 4 above)
Comment (swetharepakula) [inline pkg/backends/features/securitypolicy_test.go:81]: add tests for nil security policy, empty securityPolicy.Name

File: pkg/backends/features/securitypolicy.go [hunk 7]
```diff
@@ -31,11 +31,16 @@ import (
 // EnsureSecurityPolicy ensures the security policy link on backend service.
 // TODO(mrhohn): Emit event when attach/detach security policy to backend service.
 func EnsureSecurityPolicy(cloud *gce.Cloud, sp utils.ServicePort, be *composite.BackendService) error {
-	var desiredPolicyName string
-	if sp.BackendConfig.Spec.SecurityPolicy != nil {
-		desiredPolicyName = sp.BackendConfig.Spec.SecurityPolicy.Name
-	} else {
-		desiredPolicyName = ""
+	// Treat empty SecurityPolicy as ignored. It is too dangerous to remove user's
```
[the remark this answers is not in the document]
Comment (songrx1997) [reply pkg/backends/features/securitypolicy.go]: Done

File: pkg/backends/features/securitypolicy.go [hunk 8]
```diff
@@ -44,17 +49,14 @@ func EnsureSecurityPolicy(cloud *gce.Cloud, sp utils.ServicePort, be *composite.
 		return err
 	}
 
+	desiredPolicyName := sp.BackendConfig.Spec.SecurityPolicy.Name
```
[the remark this answers is not in the document]
Comment (songrx1997) [reply pkg/backends/features/securitypolicy.go:55]: done

File: pkg/backends/features/securitypolicy.go [hunk 9]
```diff
@@ -31,11 +31,16 @@ import (
...
+	// Treat empty SecurityPolicy as ignored. It is too dangerous to remove user's
+	// security policy that may have been configured via the UI or gcloud directly
+	// rather than via the Kubernetes API.
+	if sp.BackendConfig.Spec.SecurityPolicy == nil {
+		klog.V(2).Infof("Ignoring empty SecurityPolicy on backend service %s (%s:%s)", be.Name, sp.ID.Service, sp.ID.Port)
+		return nil
+	}
+
+	if be.Scope != meta.Global {
+		return fmt.Errorf("cloud armor security policies not supported for %s backend service %s", be.Scope, be.Name)
```
[the remark this answers is not in the document]
Comment (songrx1997) [reply pkg/backends/features/securitypolicy.go:45]: done

File: pkg/backends/features/securitypolicy.go [hunk 10]
```diff
@@ -31,11 +31,16 @@ import (
...
-	var desiredPolicyName string
-	if sp.BackendConfig.Spec.SecurityPolicy != nil {
-		desiredPolicyName = sp.BackendConfig.Spec.SecurityPolicy.Name
-	} else {
-		desiredPolicyName = ""
+	// Treat empty SecurityPolicy as ignored. It is too dangerous to remove user's
+	// security policy that may have been configured via the UI or gcloud directly
+	// rather than via the Kubernetes API.
+	if sp.BackendConfig.Spec.SecurityPolicy == nil {
+		klog.V(2).Infof("Ignoring empty SecurityPolicy on backend service %s (%s:%s)", be.Name, sp.ID.Service, sp.ID.Port)
```
[the remark this answers is not in the document]
Comment (songrx1997) [reply pkg/backends/features/securitypolicy.go]: done

File: pkg/backends/features/securitypolicy.go [hunk 11]
```diff
@@ -44,17 +49,14 @@ func EnsureSecurityPolicy(cloud *gce.Cloud, sp utils.ServicePort, be *composite.
...
 		return nil
 	}
-
-	if be.Scope != meta.Global {
-		return fmt.Errorf("cloud armor security policies not supported for %s backend service %s", be.Scope, be.Name)
-	}
-
-	klog.V(2).Infof("Set security policy in backend service %s (%s:%s) to %q", be.Name, sp.ID.Service.String(), sp.ID.Port.String(), desiredPolicyName)
+	klog.V(2).Infof("Set security policy in backend service %s (%s:%s) from %q to %q", be.Name, sp.ID.Service.String(), sp.ID.Port.String(), existingPolicyName, desiredPolicyName)
 	if err := composite.SetSecurityPolicy(cloud, be, desiredPolicyName); err != nil {
```
[the remark this answers is not in the document]
Comment (songrx1997) [reply pkg/backends/features/securitypolicy.go]: done

File: pkg/backends/features/securitypolicy.go [hunk 12]
```diff
@@ -44,17 +49,14 @@ func EnsureSecurityPolicy(cloud *gce.Cloud, sp utils.ServicePort, be *composite.
 		return err
```
[the remark this answers is not in the document]
Comment (songrx1997) [reply pkg/backends/features/securitypolicy.go:52]: done

File: pkg/backends/features/securitypolicy.go (hunk 6 above)
Comment (songrx1997) [reply pkg/backends/features/securitypolicy.go:78]: done

File: pkg/backends/features/securitypolicy_test.go (hunk 4 above)
Comment (songrx1997) [reply pkg/backends/features/securitypolicy_test.go]: done

File: pkg/backends/features/securitypolicy.go [hunk 13]
```diff
@@ -31,30 +31,49 @@ import (
...
 		return nil
 	}
 
-	if be.Scope != meta.Global {
-		return fmt.Errorf("cloud armor security policies not supported for %s backend service %s", be.Scope, be.Name)
+	if desiredPolicyName != "" {
+		klog.V(2).Infof("Set security policy in backend service %s (%s:%s) from %q to %q", be.Name, sp.ID.Service.String(), sp.ID.Port.String(), existingPolicyName, desiredPolicyName)
+		if err := composite.SetSecurityPolicy(cloud, be, desiredPolicyName); err != nil {
+			klog.V(2).Infof("Failed to set security policy from %q to %q for backend service %s (%s:%s): %v", existingPolicyName, desiredPolicyName, be.Name, sp.ID.Service.String(), sp.ID.Port.String(), err)
+			return fmt.Errorf("failed to set security policy from %q to %q for backend service %s (%s:%s): %v", existingPolicyName, desiredPolicyName, be.Name, sp.ID.Service.String(), sp.ID.Port.String(), err)
```
Comment (bowei) [inline pkg/backends/features/securitypolicy.go:66]: Usually, the way to do this is

```
err := fmt.Errorf("....")
klog.Errorf("SetSecurityPolicy() = %v", err)
return err
```

Might want to fix all the different places.

File: pkg/backends/features/securitypolicy_test.go [hunk 14]
```diff
@@ -116,10 +107,19 @@ func TestEnsureSecurityPolicy(t *testing.T) {
 			},
 		},
 		{
-			desc:                  "empty-policy",
+			desc:                  "nil-policy",
```
Comment (bowei) [inline pkg/backends/features/securitypolicy_test.go:110]: desc: "nil policy should have no effect"

Add a test case for positive and negative (e.g. policy was configured, no policy was configured).

File: pkg/metrics/features.go [hunk 15]
```diff
@@ -80,6 +80,8 @@ const (
 	// BackendConfig Features
 	cloudCDN                  = feature("CloudCDN")
 	cloudArmor                = feature("CloudArmor")
+	cloudArmorEmpty           = feature("CloudArmorEmptyString")
+	cloudArmorNone            = feature("CloudArmorNone")
```
Comment (bowei) [inline pkg/metrics/features.go:84]: None => Nil

File: pkg/metrics/features.go [hunk 16]
```diff
@@ -318,10 +320,18 @@ func featuresForServicePort(sp utils.ServicePort) []feature {
 		}
 		klog.V(6).Infof("Session affinity %s is configured for service port %s", affinityType, svcPortKey)
 	}
-	if sp.BackendConfig.Spec.SecurityPolicy != nil {
+
+	if sp.BackendConfig.Spec.SecurityPolicy == nil {
```
Comment (bowei) [inline pkg/metrics/features.go:324]: can we make this closer to the original logic 

```
if sp.BackendConfig.Spec.SecurityPolicy != nil {
    var caFeature feature
    if sp.BackendConfig.Spec.SecurityPolicy.Name == "" {
        caFeature = cloudArmorEmpty
    } else {
        caFeature = cloudArmor
    }
    features = append(features, caFeature)
		klog.V(6).Infof("Security policy %s is configured for service port %s (%s)", sp.BackendConfig.Spec.SecurityPolicy, svcPortKey, caFeature)
} else {
  features = append(features, cloudArmorNil)
		klog.V(6).Infof("Security policy %s is setting is ignored for service port %s (%s)", sp.BackendConfig.Spec.SecurityPolicy, svcPortKey)
}
```

File: pkg/backends/features/securitypolicy.go (hunk 13 above)
Comment (songrx1997) [reply pkg/backends/features/securitypolicy.go]: Done

File: pkg/backends/features/securitypolicy_test.go (hunk 14 above)
Comment (songrx1997) [reply pkg/backends/features/securitypolicy_test.go:110]: Done

File: pkg/metrics/features.go (hunk 16 above)
Comment (songrx1997) [reply pkg/metrics/features.go]: Done

File: pkg/metrics/features.go (hunk 15 above)
Comment (songrx1997) [reply pkg/metrics/features.go]: done

File: pkg/metrics/features.go [hunk 17]
```diff
@@ -80,6 +80,8 @@ const (
 	// BackendConfig Features
 	cloudCDN                  = feature("CloudCDN")
 	cloudArmor                = feature("CloudArmor")
+	cloudArmorEmpty           = feature("CloudArmorEmptyString")
+	cloudArmorNil             = feature("CloudArmorNil")
```
Comment (swetharepakula) [inline pkg/metrics/features.go:84]: add CloudArmorSet

File: pkg/metrics/features.go [hunk 18]
```diff
@@ -318,10 +320,21 @@ func featuresForServicePort(sp utils.ServicePort) []feature {
 		}
 		klog.V(6).Infof("Session affinity %s is configured for service port %s", affinityType, svcPortKey)
 	}
+
 	if sp.BackendConfig.Spec.SecurityPolicy != nil {
-		klog.V(6).Infof("Security policy %s is configured for service port %s", sp.BackendConfig.Spec.SecurityPolicy, svcPortKey)
-		features = append(features, cloudArmor)
+		var caFeature feature
```
Comment (swetharepakula) [inline pkg/metrics/features.go:325]: preserve the original metric and put the conditional logic into its own helper

File: pkg/backends/features/securitypolicy_test.go (hunk 14 above)
Comment (swetharepakula) [inline pkg/backends/features/securitypolicy_test.go:110]: we need to add a case where the currentBS has security policy set and the backend config is nil.

Comment (swetharepakula) [reply pkg/backends/features/securitypolicy_test.go:110]: this is the case below

File: pkg/backends/features/securitypolicy.go [hunk 19]
```diff
@@ -31,30 +31,53 @@ import (
...
-	if be.Scope != meta.Global {
-		return fmt.Errorf("cloud armor security policies not supported for %s backend service %s", be.Scope, be.Name)
+	if desiredPolicyName != "" {
+		klog.V(2).Infof("Set security policy in backend service %s (%s:%s) from %q to %q", be.Name, sp.ID.Service.String(), sp.ID.Port.String(), existingPolicyName, desiredPolicyName)
+		if err := composite.SetSecurityPolicy(cloud, be, desiredPolicyName); err != nil {
+			err := fmt.Errorf("failed to set security policy from %q to %q for backend service %s (%s:%s): %v", existingPolicyName, desiredPolicyName, be.Name, sp.ID.Service.String(), sp.ID.Port.String(), err)
+			klog.Errorf("SetSecurityPolicy() = %v", err)
+			return err
+		}
+		klog.V(2).Infof("Successfully set security policy in backend service %s (%s:%s) from %q to %q", be.Name, sp.ID.Service.String(), sp.ID.Port.String(), existingPolicyName, desiredPolicyName)
```
Comment (bowei) [inline pkg/backends/features/securitypolicy.go:71]: return nil after this line

File: pkg/backends/features/securitypolicy.go [hunk 20]
```diff
@@ -31,30 +31,53 @@ import (
...
 	if be.SecurityPolicy != "" && err != nil {
+		err := fmt.Errorf("failed to parse existing security policy name %q: %v", existingPolicyName, err)
+		klog.Errorf("EnsureSecurityPolicy() = %v", err)
 		return err
 	}
 
+	desiredPolicyName := sp.BackendConfig.Spec.SecurityPolicy.Name
+	klog.V(2).Infof("Current security policy: %q, desired security policy: %q", existingPolicyName, desiredPolicyName)
 	if existingPolicyName == desiredPolicyName {
+		klog.V(2).Infof("SecurityPolicy on backend service is not changed %s (%s:%s): %q", be.Name, sp.ID.Service, sp.ID.Port, sp.BackendConfig.Spec.SecurityPolicy.Name)
```
Comment (bowei) [inline pkg/backends/features/securitypolicy.go:60]: change sp.BackendConfig.Spec.SecurityPolicy.Name => desiredPolicyName

File: pkg/backends/features/securitypolicy.go [hunk 21]
```diff
@@ -31,30 +31,53 @@ import (
...
+	if desiredPolicyName != "" {
+		klog.V(2).Infof("Set security policy in backend service %s (%s:%s) from %q to %q", be.Name, sp.ID.Service.String(), sp.ID.Port.String(), existingPolicyName, desiredPolicyName)
+		if err := composite.SetSecurityPolicy(cloud, be, desiredPolicyName); err != nil {
+			err := fmt.Errorf("failed to set security policy from %q to %q for backend service %s (%s:%s): %v", existingPolicyName, desiredPolicyName, be.Name, sp.ID.Service.String(), sp.ID.Port.String(), err)
+			klog.Errorf("SetSecurityPolicy() = %v", err)
+			return err
+		}
+		klog.V(2).Infof("Successfully set security policy in backend service %s (%s:%s) from %q to %q", be.Name, sp.ID.Service.String(), sp.ID.Port.String(), existingPolicyName, desiredPolicyName)
+		return nil
+	} else {
```
Comment (bowei) [inline pkg/backends/features/securitypolicy.go:73]: you don't need else if you have a return
